clangrt builtins lib not found

Do users need to install additional library for building rocblas with DPC++ ?

-- CMAKE_BUILD_TYPE: None, set to Release by default
-- C compiler: icx was not found in PATH, using clang instead
-- The CXX compiler identification is Clang 17.0.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /path/to/dpcpp/clang++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- TARGET_DOMAINS: blas
-- Looking for dpc++
-- Performing Test is_dpcpp
-- Performing Test is_dpcpp - Success
-- Performing Test HIP_CLANG_SUPPORTS_PARALLEL_JOBS
-- Performing Test HIP_CLANG_SUPPORTS_PARALLEL_JOBS - Failed
-- Looking for C++ include pthread.h
-- Looking for C++ include pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- hip::amdhip64 is SHARED_LIBRARY
CMake Error at /opt/rocm/hip/lib/cmake/hip/hip-config.cmake:331 (message):
  clangrt builtins lib not found
Call Stack (most recent call first):
  /usr/share/cmake-3.22/Modules/CMakeFindDependencyMacro.cmake:47 (find_package)
  /opt/rocm/lib/cmake/rocblas/rocblas-config.cmake:90 (find_dependency)
  cmake/FindrocBLAS.cmake:36 (find_package)
  src/blas/backends/rocblas/CMakeLists.txt:24 (find_package)

> A fix that worked for me was either:
> 
>     * `export LIBRARY_PATH=/path/to/rocm-$rocm-version$/llvm/lib/clang/$clang-version$/lib/linux/:$LIBRARY_PATH`
>       or
> 
>     * directly modifying the `CLANGRT_BUILTINS` variable in your `CmakeCache.txt` file to point to the `libclang_rt.builtins-x86_64.a` file on your machine e.g.
>       `CLANGRT_BUILTINS:FILEPATH=/path/to/rocm-$rocm-version$/llvm/lib/clang/$clang-version$/lib/linux/libclang_rt.builtins-x86_64.a`
> 
> 
> Let me know if either of these work for you. Thanks.

@muhammad-tanvir-1211 Neither of the above steps worked resulting in the error. Where as explicitly building the `compiler-rt` external component worked though.
```
make[2]: *** No rule to make target '/lustre/orion/gen243/scratch/abagusetty/llvm/build/install/lib/clang/18/lib/linux/libclang_rt.builtins-x86_64.a', needed by 'lib/libonemkl_rng_rocrand.so.0'.  Stop.
```

--- a/cmake/FindCompiler.cmake
+++ b/cmake/FindCompiler.cmake
@@ -71,4 +71,15 @@ if(is_dpcpp)
       INTERFACE_LINK_LIBRARIES ${SYCL_LIBRARY})
   endif()
 
-endif()
+  if(ENABLE_ROCBLAS_BACKEND OR ENABLE_ROCRAND_BACKEND OR ENABLE_ROCSOLVER_BACKEND)
+    # Allow find_package(HIP) to find the correct path to libclang_rt.builtins.a
+    # HIP's CMake uses the command `${HIP_CXX_COMPILER} -print-libgcc-file-name --rtlib=compiler-rt` to find this path.
+    # This can print a non-existing file if the compiler used is icpx.
+    if(NOT HIP_CXX_COMPILER)
+      find_path(HIP_CXX_COMPILER clang++
+        HINTS ENV HIPROOT ENV ROCM_PATH
+      )
+    endif()
+  endif()
+
+endif(is_dpcpp)

--- a/docs/building_the_project_with_dpcpp.rst
+++ b/docs/building_the_project_with_dpcpp.rst
@@ -457,11 +457,12 @@ Build FAQ
 
 clangrt builtins lib not found
   Encountered when trying to build oneMKL with some ROCm libraries. There are
-  several possible solutions: * If building Open DPC++ from source, add
-  ``compiler-rt`` to the external projects compile option:
-  ``--llvm-external-projects compiler-rt``. * The *clangrt* from ROCm can be
-  used, depending on ROCm version: ``export
-  LIBRARY_PATH=/path/to/rocm-$rocm-version$/llvm/lib/clang/$clang-version$/lib/linux/:$LIBRARY_PATH``
+  several possible solutions:
+
+  * If building Open DPC++ from source, add ``compiler-rt`` to the external
+    projects compile option: ``--llvm-external-projects compiler-rt``.
+  * Manually set the variable ``HIP_CXX_COMPILER`` to HIP's toolkit ``clang++``
+    path, for instance ``-DHIP_CXX_COMPILER=/opt/rocm/6.1.0/llvm/bin/clang++``.
 
 Could NOT find CBLAS (missing: CBLAS file)
   Encountered when tests are enabled along with the BLAS domain. The tests

--- a/docs/building_the_project_with_dpcpp.rst
+++ b/docs/building_the_project_with_dpcpp.rst
@@ -463,6 +463,8 @@ clangrt builtins lib not found
     projects compile option: ``--llvm-external-projects compiler-rt``.
   * Manually set the variable ``HIP_CXX_COMPILER`` to HIP's toolkit ``clang++``
     path, for instance ``-DHIP_CXX_COMPILER=/opt/rocm/6.1.0/llvm/bin/clang++``.
+    oneMKL may fail to link if the clang versions of ``icpx`` and ``rocm`` are
+    not compatible.
 
 Could NOT find CBLAS (missing: CBLAS file)
   Encountered when tests are enabled along with the BLAS domain. The tests

